task numbat {
    File count_mtrx
    File allele_df
    File reference_mtrx
    File reference_annotation
    String genome_build
    Boolean draw_plots = true
    # runtime
    Int ncores = 8

    command {
        R --no-save  <<RSCRIPT
        library(numbat)
        # TODO: improve this by using referen_annotation to separate reference and case count matrices
        count_mtrx = readRDS("~{count_mtrx}")
        allele_df = read.table(gzfile("~{allele_df}"), sep='\t', header=TRUE)
        reference_mtrx = readRDS("~{reference_mtrx}")
        ref_annot = read.table("~{reference_annotation}", sep=",", header=TRUE) #fix this

        # aggregate reference counts
        ref_internal = aggregate_counts(reference_mtrx, ref_annot, normalized=TRUE)
        out = run_numbat(
            count_mtrx, # gene x cell integer UMI count matrix 
            ref_internal, # reference expression profile, a gene x cell type normalized expression level matrix
            allele_df, # allele dataframe generated by pileup_and_phase script
            genome = ~{genome_build},
            t = 1e-5,
            ncores = ~{ncores},
            plot = ~{draw_plots},
            out_dir = './results/'
        )
        RSCRIPT
    }

    output {
        Array[File] numbat_results = glob("results/*")
    }
}